common/lib/dependabot/file_parsers/base/dependency_set.rb

```diff
@@ -50,41 +60,101 @@ def case_sensitive?
           @case_sensitive
         end
 
-        def dependency_for_name(name)
-          return dependencies.find { |d| d.name == name } if case_sensitive?
+        def key_for_name(name)
+          case_sensitive? ? name : name.downcase
+        end
 
-          dependencies.find { |d| d.name&.downcase == name&.downcase }
+        def key_for(dep)
+          key_for_name(dep.name)
         end
 
-        def combined_dependency(old_dep, new_dep)
-          package_manager = old_dep.package_manager
-          v_cls = Utils.version_class_for_package_manager(package_manager)
-
-          # If we already have a requirement use the existing version
-          # (if present). Otherwise, use whatever the lowest version is
-          new_version =
-            if old_dep.requirements.any? then old_dep.version || new_dep.version
-            elsif !v_cls.correct?(new_dep.version) then old_dep.version
-            elsif !v_cls.correct?(old_dep.version) then new_dep.version
-            elsif v_cls.new(new_dep.version) > v_cls.new(old_dep.version)
-              old_dep.version
+        class DependencySlot
```

Comment:
Can we add a short bit of documentation about what a `DependencySlot` is here? From my understanding right now something like:

```ruby
# DependencySlot represents a single Dependency with one or more versions of that dependency.
# This exists for ecosystems (like npm_and_yarn) where it's possible to pull in multiple versions of
# a dependency in a project. This way we can keep track of all those versions while performing an
# update.
```

Reply:
Also could be in it's own file, but I don't feel super strongly about that

Reply:
Documentation added in ab9036b3cff2294b0f5306f4e4f563aa77317c87